rp2040-hal/src/timer.rs

```diff
@@ -84,27 +94,85 @@ impl Timer {
     }
     /// Retrieve a reference to alarm 0. Will only return a value the first time this is called
     pub fn alarm_0(&mut self) -> Option<Alarm0> {
-        take_alarm(1 << 0).then_some(Alarm0(PhantomData))
+        take_alarm(1 << 0).then_some(Alarm0(*self))
     }
 
     /// Retrieve a reference to alarm 1. Will only return a value the first time this is called
     pub fn alarm_1(&mut self) -> Option<Alarm1> {
-        take_alarm(1 << 1).then_some(Alarm1(PhantomData))
+        take_alarm(1 << 1).then_some(Alarm1(*self))
     }
 
     /// Retrieve a reference to alarm 2. Will only return a value the first time this is called
     pub fn alarm_2(&mut self) -> Option<Alarm2> {
-        take_alarm(1 << 2).then_some(Alarm2(PhantomData))
+        take_alarm(1 << 2).then_some(Alarm2(*self))
     }
 
     /// Retrieve a reference to alarm 3. Will only return a value the first time this is called
     pub fn alarm_3(&mut self) -> Option<Alarm3> {
-        take_alarm(1 << 3).then_some(Alarm3(PhantomData))
+        take_alarm(1 << 3).then_some(Alarm3(*self))
+    }
+
+    /// Pauses execution for at minimum `us` microseconds.
+    fn delay_us(&self, mut us: u32) {
+        let mut start = self.get_counter_low();
+        // If we knew that the loop ran at least once per timer tick,
+        // this could be simplified to:
+        // ```
+        // while timer.timelr.read().bits().wrapping_sub(start) <= us {
+        //     cortex_m::asm::nop();
+        // }
+        // ```
+        // However, due to interrupts, for `us == u32::MAX`, we could
+        // miss the moment where the loop should terminate if the loop skips
+        // a timer tick.
+        loop {
+            let now = self.get_counter_low();
+            let waited = now.wrapping_sub(start);
+            if waited >= us {
+                break;
+            }
+            start = now;
+            us -= waited;
+        }
     }
 }
 
-// safety: all write operations are synchronised and all reads are atomic
-unsafe impl Sync for Timer {}
+macro_rules! impl_delay_traits {
+    ($($t:ty),+) => {
+        $(
+        impl embedded_hal::blocking::delay::DelayUs<$t> for Timer {
+            fn delay_us(&mut self, mut us: $t) {
+                #![allow(unused_comparisons)]
+                assert!(us >= 0); // Only meaningful for i32
```

Comment:
Is this completely elliminated when the comparison is unused ?

Reply:
I think so, yes. I checked the generated assembly (at optimizer level "z") for u16, u32 and u64, as well as i32 with a non-negative constant value. In all those cases the comparison with 0 and the branch with the panic call get removed by the compiler.